pkg/codec/hard_coder.go

```diff
@@ -0,0 +1,107 @@
+package codec
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+)
+
+func NewHardCoder(onChain map[string]any, offChain map[string]any, hooks ...mapstructure.DecodeHookFunc) (Modifier, error) {
+	if err := verifyHardCodeKeys(onChain); err != nil {
+		return nil, err
+	} else if err = verifyHardCodeKeys(offChain); err != nil {
+		return nil, err
+	}
+
+	myHooks := make([]mapstructure.DecodeHookFunc, len(hooks)+1)
+	copy(myHooks, hooks)
+	myHooks[len(hooks)] = hardCodeManyHook
+
+	m := &onChainHardCoder{
+		modifierBase: modifierBase[any]{
+			fields:           offChain,
+			onToOffChainType: map[reflect.Type]reflect.Type{},
+			offToOnChainType: map[reflect.Type]reflect.Type{},
+		},
+		onChain: onChain,
+		hooks:   myHooks,
+	}
+	m.modifyFieldForInput = func(field *reflect.StructField, key string, v any) error {
+		// if we are typing it differently, we need to make sure it's hard-coded the other way
+		newType := reflect.TypeOf(v)
+		if _, ok := m.onChain[key]; !ok && field.Type != newType {
+			return fmt.Errorf(
+				"%w: cannot change field type without hard-coding its onchain value for key %s",
+				types.ErrInvalidType,
+				key)
+		}
+		field.Type = newType
+		return nil
+	}
+	m.addFieldForInput = func(key string, value any) reflect.StructField {
+		return reflect.StructField{
+			Name: key,
+			Type: reflect.TypeOf(value),
+		}
+	}
+	return m, nil
+}
+
+type onChainHardCoder struct {
+	modifierBase[any]
+	onChain map[string]any
+	hooks   []mapstructure.DecodeHookFunc
+}
+
+func verifyHardCodeKeys(values map[string]any) error {
+	seen := map[string]bool{}
+	for _, k := range subkeysLast(values) {
+		parts := strings.Split(k, ".")
+		on := ""
+		for _, part := range parts {
+			on += part
+			if seen[on] {
+				return fmt.Errorf("%w: key %s and %s cannot both be present", types.ErrInvalidConfig, on, k)
+			}
+		}
+		seen[k] = true
+	}
+	return nil
+}
+
+func (o *onChainHardCoder) TransformForOnChain(offChainValue any) (any, error) {
+	return transformWithMaps(offChainValue, o.offToOnChainType, o.onChain, hardCode, o.hooks...)
+}
+
+func (o *onChainHardCoder) TransformForOffChain(onChainValue any) (any, error) {
+	return transformWithMaps(onChainValue, o.onToOffChainType, o.fields, hardCode, hardCodeManyHook)
+}
+
+func hardCode(extractMap map[string]any, key string, item any) error {
+	extractMap[key] = item
+	return nil
+}
+
+func hardCodeManyHook(from reflect.Value, to reflect.Value) (any, error) {
+	switch to.Kind() {
+	case reflect.Slice, reflect.Array:
+		switch from.Kind() {
+		case reflect.Slice, reflect.Array:
+			return from.Interface(), nil
```

Comment:
There's a different mapstructure hook for that when decoding in utils.

This one just makes it so that if you say A.B = 10 and A is a slice all A.B will be 10.  I'll add a comment to make it more clear.